pennylane/transforms/batch_transform.py

```diff
@@ -0,0 +1,258 @@
+    r"""Class for registering a tape transform that takes a tape, and outputs
+    a batch of tapes to be independently executed on a quantum device.
+
+    Examples of such transforms include quantum gradient shift rules (such
+    as finite-differences and the parameter-shift rule) and metrics such as
+    the quantum Fisher information matrix.
+
+    Args:
+        transform_fn (function): The function to register as the batch tape transform.
+            It can have an arbitrary number of arguments, but the first argument
+            **must** be the input tape.
+        expand_fn (function): An expansion function (if required) to be applied to the
+            input tape before the transformation takes place.
+        differentiable (bool): Specifies whether the transform is differentiable or
+            not. A transform may be non-differentiable for several reasons:
+
+            - It does not use an autodiff framework for its tensor manipulations;
+            - It returns a non-differentiable or non-numeric quantity, such as
+              a boolean, string, or integer.
+
+            In such a case, setting ``differentiable=False`` instructs the decorator
+            to mark the output as 'constant', reducing potential overhead.
+
+    **Example**
+
+    A valid batch tape transform is a function that satisfies the following:
+
+    - The first argument must be a tape.
+
+    - Depending on the structure of this input tape, various quantum operations, functions,
+      and templates may be called.
+
+    - Any internal classical processing should use the ``qml.math`` module to ensure
+      the transform is differentiable.
+
+    - The transform should return a tuple containing:
+
+      * Multiple transformed tapes to be executed on a device.
+      * A classical processing function for post-processing the executed tape results.
+        This processing function should have the signature ``f(list[tensor_like]) → Any``.
+        If ``None``, no classical processing is applied to the results.
+
+    For example:
+
+    .. code-block:: python
+
+        @qml.batch_transform
+        def my_transform(tape, a, b):
+            "Generates two tapes, one with all RX replaced with RY,
+            and the other with all RX replaced with RZ."
+
+            tape1 = qml.tape.JacobianTape()
+            tape2 = qml.tape.JacobianTape()
+
+            # loop through all operations on the input tape
+            for op in tape.operations + tape.measurements:
+                if op.name == "RX":
+                    wires = op.wires
+                    param = op.parameters[0]
+
+                    with tape1:
+                        qml.RY(a * qml.math.abs(param), wires=wires)
+
+                    with tape2:
+                        qml.RZ(b * qml.math.abs(param), wires=wires)
+                else:
+                    for t in [tape1, tape2]:
+                        with t:
+                            qml.apply(op)

```

Comment:
Not related to this PR, but would be cool for these cases to be able to do

```
qml.apply(op, context=[tape1, tape2])
```

rather than having to loop over tapes.

Reply:
I know 😿 I actually tried this originally:

```python
with tape1, tape2:
    qml.apply(op)
```
thinking it would apply to both. But it doesn't 😢